tracing.go

```diff
@@ -84,53 +94,167 @@ func (s Span) ParentID() uint64 { return s.parentID }
+func CurrentSpan(ctx context.Context) *Span {
+	if sp := opentracing.SpanFromContext(ctx); sp != nil {
+		var injectable injectableSpan
+		if err := injectable.initFromOpenTracing(sp); err == nil {
+			span := Span(injectable)
+			return &span
+		}
+		// return empty span on error, instead of possibly a partially filled one
+	}
+	return &emptySpan
+}
 
-// NewChildSpan begins a new child span in the provided Context
-func (s Span) NewChildSpan() *Span {
-	childSpan := &Span{
-		traceID:  s.traceID,
-		parentID: s.spanID,
-		flags:    s.flags,
+// startOutboundSpan creates a new tracing span to represent the outbound RPC call.
+// If the context already contains a span, it will be used as a parent, otherwise
+// a new root span is created.
+//
+// If the tracer supports Zipkin-style trace IDs, then call.callReq.Tracing is
+// initialized with those IDs. Otherwise it is assigned random values.
+func (c *Connection) startOutboundSpan(ctx context.Context, serviceName, methodName string, call *OutboundCall, startTime time.Time) opentracing.Span {
+	var parent opentracing.SpanContext // ok to be nil
+	if s := opentracing.SpanFromContext(ctx); s != nil {
+		parent = s.Context()
+	}
+	span := c.Tracer().StartSpan(
+		methodName,
+		opentracing.ChildOf(parent),
+		opentracing.StartTime(startTime),
+	)
+	if isTracingDisabled(ctx) {
+		ext.SamplingPriority.Set(span, 0)
 	}
-	if s.spanID == 0 {
-		childSpan.spanID = childSpan.traceID
+	ext.SpanKindRPCClient.Set(span)
+	ext.PeerService.Set(span, serviceName)
+	setPeerHostPort(span, c.remotePeerInfo.HostPort)
+	span.SetTag("as", call.callReq.Headers[ArgScheme])
+	injectable := injectableSpan{}
+	if err := injectable.initFromOpenTracing(span); err == nil {
+		call.callReq.Tracing = Span(injectable)
 	} else {
-		childSpan.spanID = uint64(traceRng.Int63())
+		call.callReq.Tracing.initRandom()
 	}
-	return childSpan
+	return span
 }
 
-func (s *Span) sampleRootSpan(sampleRate float64) {
-	// Sampling only affects root spans
-	if s.ParentID() != 0 {
-		return
+// InjectOutboundSpan retrieves OpenTracing Span from `response`, where it is stored
+// when the outbound call is initiated. The tracing API is used to serialize the span
+// into the application `headers`, which will propagate tracing context to the server.
+// Returns modified headers containing serialized tracing context.
+func InjectOutboundSpan(response *OutboundCallResponse, headers map[string]string) map[string]string {
+	span := response.span
+	if span == nil {
+		return headers
+	}
+	if headers == nil {
+		headers = make(map[string]string)
 	}
+	carrier := tracingHeadersCarrier(headers)
+	if err := span.Tracer().Inject(span.Context(), opentracing.TextMap, carrier); err != nil {
+		// Something had to go seriously wrong for Inject to fail, usually a setup problem.
+		// A good Tracer implementation may also emit a metric.
+		response.log.Error("Failed to inject tracing span: " + err.Error())
```

Comment:
can you use structured fields for errors, `log.WithFields(...).Error("Failed to inject tracing span.")` (the log message is constant and a full sentence ending with a period)